### PR TITLE
net: lwm2m: response to peer with correct error code when write fail

### PR DESCRIPTION
**Pull request by rbtchc**
Update the firmware update_result accordingly by checking return
value of the firmware data writes callback registered by the application.
Also, set response code according.

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c [hunk 1]
```diff
@@ -249,8 +249,9 @@ do_firmware_transfer_reply_cb(const struct coap_packet *response,
 		SYS_LOG_ERR("Unexpected response from server: %d.%d",
 			    COAP_RESPONSE_CODE_CLASS(resp_code),
 			    COAP_RESPONSE_CODE_DETAIL(resp_code));
-		lwm2m_firmware_set_update_result(RESULT_CONNECTION_LOST);
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: We used to set result to RESULT_CONNECTION_LOST here.  Now it's -ENOENT which isn't handled directly below so it falls back to RESULT_UPDATE_FAILED.

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware.c [hunk 2]
```diff
@@ -196,19 +196,27 @@ static int package_write_cb(u16_t obj_inst_id,
...
+		if (last_block) {
+			lwm2m_firmware_set_update_state(STATE_DOWNLOADED);
 		}
+
+		return 0;
 	}
 
-	if (last_block) {
-		lwm2m_firmware_set_update_state(STATE_DOWNLOADED);
+	if (ret == -ENOMEM) {
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware.c]: make this a continuation of the check above with "} else if (ret == -ENOMEM) {"?

**Review by mike-scott — CHANGES_REQUESTED** [review]

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c (hunk 1 above)
Comment (rbtchc) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: I'm going to use "-ENOMSG /* Unexpected message type */" to represent the case.
ENOENT seems a common error number and we might have other issues if we map it to connection lost.
Or maybe we just set the state and return here directly.

**Comment by rbtchc** [comment]
Update according to @mike-scott 's comment

https://hastebin.com/esayeyamay.lua

**Review by mike-scott — APPROVED** [review]
LGTM

**Review by jukkar — APPROVED** [review]

**Pull request merged**